lib/elixir/lib/application.ex

```diff
@@ -734,9 +738,19 @@ defmodule Application do
   """
   @spec delete_env(app, key, timeout: timeout, persistent: boolean) :: :ok
   def delete_env(app, key, opts \\ []) when is_atom(app) do
+    maybe_warn_on_app_env_key(app, key)
     :application.unset_env(app, key, opts)
   end
 
+  # TODO: turn warning into error on Elixir v2.0
+  defp maybe_warn_on_app_env_key(_app, key) when is_atom(key),
+    do: :ok
+
+  defp maybe_warn_on_app_env_key(app, key) do
+    message = "passing non-atom as application env key is deprecated (got: `#{inspect(key)}`)"
```

Comment:
We don't use markdown in error messages :) My suggestion:

```suggestion
    message = "passing non-atom as application env key is deprecated, got: #{inspect(key)}"
```